src/coreclr/src/jit/codegenlinear.cpp

```diff
@@ -985,99 +1064,26 @@ void CodeGen::genUnspillRegIfNeeded(GenTree* tree)
 #endif
-
-            // TODO-Review: We would like to call:
-            //      genUpdateRegLife(varDsc, /*isBorn*/ true, /*isDying*/ false DEBUGARG(tree));
-            // instead of the following code, but this ends up hitting this assert:
-            //      assert((regSet.GetMaskVars() & regMask) == 0);
-            // due to issues with LSRA resolution moves.
-            // So, just force it for now. This probably indicates a condition that creates a GC hole!
-            //
-            // Extra note: I think we really want to call something like gcInfo.gcUpdateForRegVarMove,
-            // because the variable is not really going live or dead, but that method is somewhat poorly
-            // factored because it, in turn, updates rsMaskVars which is part of RegSet not GCInfo.
-            // TODO-Cleanup: This code exists in other CodeGen*.cpp files, and should be moved to CodeGenCommon.cpp.
-
-            // Don't update the variable's location if we are just re-spilling it again.
-
-            if ((unspillTree->gtFlags & GTF_SPILL) == 0)
-            {
-                genUpdateVarReg(varDsc, tree);
-
-#ifdef USING_VARIABLE_LIVE_RANGE
-                // We want "VariableLiveRange" inclusive on the beginbing and exclusive on the ending.
-                // For that we shouldn't report an update of the variable location if is becoming dead
-                // on the same native offset.
-                if ((unspillTree->gtFlags & GTF_VAR_DEATH) == 0)
-                {
-                    // Report the home change for this variable
-                    varLiveKeeper->siUpdateVariableLiveRange(varDsc, lcl->GetLclNum());
-                }
-#endif // USING_VARIABLE_LIVE_RANGE
-
-                if (!varDsc->lvLiveInOutOfHndlr)
-                {
-#ifdef DEBUG
-                    if (VarSetOps::IsMember(compiler, gcInfo.gcVarPtrSetCur, varDsc->lvVarIndex))
-                    {
-                        JITDUMP("\t\t\t\t\t\t\tRemoving V%02u from gcVarPtrSetCur\n", lcl->GetLclNum());
-                    }
-#endif // DEBUG
-                    VarSetOps::RemoveElemD(compiler, gcInfo.gcVarPtrSetCur, varDsc->lvVarIndex);
-                }
-
-#ifdef DEBUG
-                if (compiler->verbose)
-                {
-                    printf("\t\t\t\t\t\t\tV%02u in reg ", lcl->GetLclNum());
-                    varDsc->PrintVarReg();
-                    printf(" is becoming live  ");
-                    compiler->printTreeID(unspillTree);
-                    printf("\n");
-                }
-#endif // DEBUG
-
-                regSet.AddMaskVars(genGetRegMask(varDsc));
-            }
-
-            gcInfo.gcMarkRegPtrVal(dstReg, unspillTree->TypeGet());
+            bool reSpill   = ((unspillTree->gtFlags & GTF_SPILL) != 0);
+            bool isLastUse = lcl->IsLastUse(0);
+            genUnspillLocal(lcl->GetLclNum(), spillType, lcl->AsLclVar(), dstReg, reSpill, isLastUse);
```

Comment:
Nit1: I think we probably do not need to pass both `lcl->GetLclNum()` and `lcl->AsLclVar()`.
Nit2: `lcl` is already a `LclVar`.

Reply:
> I think we probably do not need to pass both lcl->GetLclNum() and lcl->AsLclVar()

We need both because when `lcl` is a multi-reg node we need to pass the node as well as the actual varnum of the specific field that's being spilled.

Reply:
Hm, I am missing something,
```
void CodeGen::genUnspillLocal(GenTreeLclVar* lclNode, var_types type, regNumber regNum, bool reSpill, bool isLastUse)
{
  unsigned varNum = lcl->GetLclNum();
```

won't it work?

Reply:
A multi-reg local node will have the varNum of the parent. Each field has dedicated flags for last-use, spill and unspill, and they are in order of the fields' varNums (starting with `lvFieldLclStart`). For now (with no multi-reg nodes) these will be the same.

Reply:
Got it, thank you!